src/operators.jl

```diff
@@ -363,201 +363,163 @@ function Compat.LinearAlgebra.diagm(x::AbstractVector{<:AbstractVariableRef})
     diagm(0=>copyto!(similar(x, GenericAffExpr{Float64,eltype(x)}), x))
 end
 
-###############
-# The _multiply!(buf,y,z) adds the results of y*z into the buffer buf. No bounds/size
-# checks are performed; it is expected that the caller has done this, has ensured
-# that the eltype of buf is appropriate, and has zeroed the elements of buf (if desired).
-
-function _multiply!(ret::Array{T}, lhs::Array, rhs::Array) where T<:JuMPTypes
-    m, n = size(lhs,1), size(lhs,2)
-    r, s = size(rhs,1), size(rhs,2)
-    for i ∈ 1:m, j ∈ 1:s
-        q = ret[i,j]
-        _sizehint_expr!(q, n)
-        for k ∈ 1:n
-            tmp = convert(T, lhs[i,k]*rhs[k,j])
-            add_to_expression!(q, tmp)
-        end
+###############################################################################
+# Matrix/Vector Arithmetic with JuMP eltypes
+###############################################################################
+
+###############################################################################
+# convenience/utility definitions
+
+const GenericAffOrQuadExpr = Union{GenericAffExpr,GenericQuadExpr}
+
+densify_with_jump_eltype(x::AbstractMatrix) = convert(Matrix, x)
+
+function densify_with_jump_eltype(x::SparseMatrixCSC{V}) where V<:AbstractVariableRef
```

Comment:
Just because (AFAICT, maybe I missed something) that was what the code was doing before, when it was pirate-overloading `Matrix`. I could change to `AbstractMatrix{<:AbstractVariableRef}` if you want.